R/pmcmc_parameters.R

```diff
@@ -132,50 +134,142 @@ pmcmc_parameters <- R6::R6Class(
 
-      if (is.null(transform)) {
-        transform <- as.list
-      }
-      assert_is(transform, "function")
+      # Split constructor for different proposal types. First asssumes array
+      #  with 3 dimensions, second is matrix.
+      varied <- vlapply(parameters, inherits, what = "pmcmc_varied_parameter")
+      if (any(varied)) {
+        # is.null(private$varied) quick check to see if all fixed
+        private$varied <- vcapply(parameters[varied], "[[", "name")
+        private$populations <- assert_character(populations)
+        lapply(parameters, function(.x) {
+          if (!(length(.x$initial) %in% c(1, length(populations)))) {
+            stop(sprintf(
+              "Expected length '1' or '%d' but got '%d' for parameter '%s'.",
+              length(populations), length(.x$initial), .x$name
+            ))
+          }
+        })
 
-      assert_is(proposal, "matrix")
-      if (!all(dim(proposal) == length(parameters))) {
-        stop(sprintf(
-          "Expected a square proposal matrix with %d rows and columns",
-          length(parameters)))
-      }
-      if (!is.null(dimnames(proposal))) {
-        ## At this point we could reorder if that's useful and these
-        ## are setequal.
-        ok <- identical(rownames(proposal), names(parameters)) &&
-          identical(colnames(proposal), names(parameters))
-        if (!ok) {
-          stop("Expected dimension names of 'proposal' to match parmeters")
+        # proposal
+        assert_is(proposal, "array")
+        dims <- dim(proposal)
+
+        # assume that if matrix provided then same proposal for all populations
+        if (is.na(dims[3])) {
+          arr_proposal <- array(proposal,
+            dim = c(dim(proposal), length(populations)))
+          if (!is.null(dimnames(proposal))) {
+            dimnames(arr_proposal) <- c(dimnames(proposal), list(populations))
+          }
+          proposal <- arr_proposal
+          dims <- dim(proposal)
+        }
+
+        # catch wrong dimension sizes
+        if (dims[3] != length(populations) ||
+            !all(dims[1:2] == length(parameters))) {
+          stop(sprintf(
+            "Expected proposal array with dimensions %d x %d x %d.",
+            length(parameters), length(parameters), length(populations)))
         }
+
+        if (!is.null(dimnames(proposal))) {
+          ## At this point we could reorder if that's useful and these
+          ## are setequal.
+          ok <- identical(rownames(proposal), names(parameters)) &&
+            identical(colnames(proposal), names(parameters)) &&
+            identical(dimnames(proposal)[3], populations)
+          if (!ok) {
+            stop("Expected dimension names of 'proposal' to match parameters
+            and populations.")
+          }
+        }
+
+         private$proposal <- apply(proposal, 3, rmvnorm_generator)
+         private$min <- sapply(parameters, function(.x) {
```

Comment:
Please don't use `sapply`

The function itself can be simplified given the checks above
```r
private$min <- vapply(parameters, function(x) rep_len(x, length(populations)),
                      numeric(length(populations)))
```